VIEW: import

jepett0 opened the pull request:
### Enable views in imports

View imports go through these steps:
- GetScheme, download the scheme from S3 (this stage is the same for views and tables)
- modify the query a bit (search for the [RewriteCreateViewQuery](https://github.com/ydb-platform/ydb/blob/4b0114f60331a2949089181b15fb4bccafd1147c/ydb/core/tx/schemeshard/schemeshard_import__create.cpp#L813) function call in the OnSchemeResult function)
- CreateTable, (should have been named CreateView, but we don't want to create a separate step in the import pipeline just for the sake of a better naming), **prepares** the create view query that was downloaded and patched in the previous steps
- set the PreparedCreationQuery of the view item (note: this field [is persisted](https://github.com/ydb-platform/ydb/pull/13361/files#diff-00ea23847aa192a17104a39b27adfb59c772fddbcf025d982f8605b243ede114))
- ExecutePreparedQuery, see the corresponding function [here](https://github.com/ydb-platform/ydb/blob/4b0114f60331a2949089181b15fb4bccafd1147c/ydb/core/tx/schemeshard/schemeshard_import__create.cpp#L354). It is important to notice that we add permissions that were stored in the S3 to the modify scheme event at this step.
- OnModifyResult, same as for the tables
- OnNotifyResult, same as for the tables (view items [go](https://github.com/ydb-platform/ydb/blob/4b0114f60331a2949089181b15fb4bccafd1147c/ydb/core/tx/schemeshard/schemeshard_import__create.cpp#L1141) straight to the `Done` state from here)

In terms of the import [EStates](https://github.com/ydb-platform/ydb/blob/83ee3897afc43f04ad506abd6838df5dc4725444/ydb/core/tx/schemeshard/schemeshard_info_types.h#L2821) the view item progression is the following:
- GetScheme
- CreateTable
  - PreparedCreationQuery is **not** set
  - PreparedCreationQuery is set
- Done

View creation is retried in case of scheme errors (scheme error is the error code that is returned in the case (among others) when the view references a table that does not exist). There is only 1 retry per view.

The main commit is:
- https://github.com/ydb-platform/ydb/pull/13361/commits/32c7f7c9f60cfd5334ed1b6a6210f3e52f026772

A couple of bug fixes along the way (please verify, if you know the code):
- https://github.com/ydb-platform/ydb/pull/13361/commits/ae88674773716ad9f6fface57b4c55953e7a15d1
- https://github.com/ydb-platform/ydb/pull/13361/commits/3030a58954574631a3bd5141ae1ce75a74e86600

CyberROFL commented on ydb/core/tx/schemeshard/schemeshard_impl.cpp:
```diff
@@ -4551,6 +4551,15 @@ void TSchemeShard::Die(const TActorContext &ctx) {
     if (TabletMigrator) {
         ctx.Send(TabletMigrator, new TEvents::TEvPoisonPill());
     }
+    for (const auto& [id, exportInfo] : Exports) {
+        if (!exportInfo->IsDone()) {
+            for (const auto& item : exportInfo->Items) {
+                if (item.SchemeUploader != TActorId()) {
+                    ctx.Send(item.SchemeUploader, new TEvents::TEvPoisonPill());
+                }
+            }
+        }
+    }
```
Всё-таки, лучше положить это в отдельный контейнер, типа` THashMap<std::pair<ui64, ui32>, TActorId>`, чтобы в нем лежали только активные actorid, которые нужно прибить в случае рестарта.

CyberROFL commented on ydb/core/tx/schemeshard/schemeshard_private.h:
```diff
@@ -95,6 +101,39 @@ namespace TEvPrivate {
+    };
+
+    struct TEvExportSchemeUploadResult: public TEventLocal<TEvExportSchemeUploadResult, EvExportSchemeUploadResult> {
+        ui64 ExportId;
+        ui32 ItemIdx;
+        bool Success;
+        TString Error;
```
А тут почему не `const`?

jepett0 replied:
никак не выберу на какой стул сесть:

1) везде, где можно писать const и получить, пожалуй, слишком визуально перегруженный и к тому же сильно отличающийся по стилю от типичного для YDB код
2) писать без const, когда value, а const добавлять только для ссылок, потому что для них эта разница действительно важна

В этом месте добавлю const, а то от соседей отличается

CyberROFL commented on ydb/core/tx/schemeshard/schemeshard_schema.h:
```diff
@@ -1553,6 +1555,7 @@ struct Schema : NIceDb::Schema {
         struct Scheme : Column<6, NScheme::NTypeIds::String> {};
         struct Permissions : Column<11, NScheme::NTypeIds::String> {};
         struct Metadata : Column<12, NScheme::NTypeIds::String> {};
+        struct CreationQuery : Column<13, NScheme::NTypeIds::String> {};
```
Наверно, это должно быть `Utf8`?

CyberROFL commented on ydb/core/tx/schemeshard/schemeshard_export__create.cpp:
```diff
@@ -311,7 +327,23 @@ struct TSchemeShard::TExport::TTxProgress: public TSchemeShard::TXxport::TTxBase
+        } else if (item.SourcePathType == NKikimrSchemeOp::EPathTypeView) {
+            Ydb::Export::ExportToS3Settings exportSettings;
+            Y_ABORT_UNLESS(exportSettings.ParseFromString(exportInfo->Settings));
+            const auto databaseRoot = TStringBuilder() << '/' << JoinSeq('/', Self->RootPathElements);
```
Используй `CanonizePath`.

jepett0 replied:
А мы же здесь полностью контролируем результат, нет никакого пользовательского input'а. Может, руками лучше собирать? Разве здесь не каноничный путь может получится?

jepett0 replied:
ok. Применил правку, предложенную в отдельном PR для export'а:

https://github.com/ydb-platform/ydb/pull/13677#discussion_r1925098088

CyberROFL commented on ydb/core/tx/schemeshard/schemeshard_export__create.cpp:
```diff
@@ -311,7 +327,23 @@ struct TSchemeShard::TExport::TTxProgress: public TSchemeShard::TXxport::TTxBase
+            const auto databaseRoot = TStringBuilder() << '/' << JoinSeq('/', Self->RootPathElements);
+
+            NBackup::TMetadata metadata;
+            // do not check hashes of views in backups
```
Напиши TODO, что это временно.

CyberROFL commented on ydb/core/tx/schemeshard/schemeshard_export_scheme_uploader.cpp:
```diff
@@ -0,0 +1,362 @@
+            return;
+        }
+        if (!Scheme || !Permissions) {
+            Send(SchemeShard, new TEvSchemeShard::TEvDescribeScheme(SourcePathId));
+            this->Become(&TThis::StateDescribe);
+            return;
```
Если сделать функцию `void Describe()`, которая отправит запрос и переключит состояние, а тут написать `return Describe()`, то будет красивее, типа такого:
```cpp
if (!Scheme || !Permissions) {
    return Describe();
}
if (!SchemeUploaded) {
    return UploadScheme();
}
...
```

CyberROFL commented on ydb/core/tx/schemeshard/schemeshard_export_scheme_uploader.cpp:
```diff
@@ -0,0 +1,362 @@
+        auto reportError = [&](const TString& error) {
+            Send(SchemeShard, new TEvPrivate::TEvExportSchemeUploadResult(
+                ExportId, ItemIdx, false, error
+            ));
```
Если после этого актор ничего не делает, то он должен завершиться, т.е. вызвать `PassAway()`.
А лучше заменить вызов `reportError` на `Finish(false, error)` — там уже делается всё, что нужно.

jepett0 replied:
Проворонил. Спасибо, что заметил!

jepett0 committed:
view import: preliminary refactoring

jepett0 committed:
enable view import

- server side view import code

jepett0 committed:
reboots: tests with views

jepett0 committed:
reboots: refactoring to enable KQP and test view import

jepett0 committed:
OnModifyResult bugfix for already existing paths

jepett0 committed:
reboots: bugfixes in cancel test suite

- We have been waiting until the import finishes before cancelling it. The status of the import would always be "success" in this case.
- Check that if the status is "cancelled", then it was cancelled manually and not due to some other unexpected error.

CyberROFL reviewed (commented):
lgtm

jepett0 committed:
minor refactoring

jepett0 committed:
export: refactoring for greater similarity in child actor treatment

jepett0 committed:
import: remember and kill child actors

jepett0 committed:
review fixes

pixcc commented on ydb/core/tx/schemeshard/ut_restore/ut_restore.cpp:
```diff
@@ -4559,36 +4613,96 @@ Y_UNIT_TEST_SUITE(TImportWithRebootsTests) {
+                        EPathTypeView,
+                        R"(
+                            -- backup root: "/MyRoot"
+                            CREATE VIEW IF NOT EXISTS `view` WITH security_invoker = TRUE AS SELECT 1;
```
Нужен тест на импорт, где есть вью и таблица. Вью ссылается на таблицу.

pixcc commented on ydb/core/tx/schemeshard/schemeshard_import__create.cpp:
```diff
@@ -338,6 +351,65 @@ struct TSchemeShard::TImport::TTxProgress: public TSchemeShard::TXxport::TTxBase
+        TVector<ui32> retriedItems;
+        for (ui32 itemIdx : xrange(importInfo->Items.size())) {
+            auto& item = importInfo->Items[itemIdx];
+            if (!item.CreationQuery.empty() && item.ViewCreationRetries == 0) {
```
Ты тут проходишься по всем айтемам и ретраишь только те, которые до этого не ретраились, в независимости от их стейта (может быть и не нужно их ретраить, потому что все успешно создалось). Больше ты их ретраить не будешь, потому что внутри цикла инкрементируешь счетчик, но одного ретрая может не хватить для создания вью.

jepett0 replied:
Добавил проверку на то, что item ждёт (этот state использую только я и он означает, что создание этого view уже один раз пофейлилось со статусом scheme error).

Проблему нескольких ретраев будем решать в отдельном тикете:
- https://github.com/ydb-platform/ydb/issues/13932

jepett0 committed:
IsCreatedByQuery refactoring

jepett0 commented on ydb/core/tx/schemeshard/schemeshard_import__create.cpp:
```diff
@@ -970,6 +1138,10 @@ struct TSchemeShard::TImport::TTxProgress: public TSchemeShard::TXxport::TTxBase
 
         switch (item.State) {
         case EState::CreateTable:
```
Должно быть просто имя enum'а, которое ни на что не влияет и чьё изменение не сломает обратной совместимости. Может, конечно, где-то в логах проскакивает это название, но я не нашёл. Поменяю на CreateSchemeObject

jepett0 committed:
review fix: rename CreateTable import state to CreateSchemeObject

jepett0 committed:
test view creation retries on import

jepett0 committed:
add a check that the item is waiting on retries

jepett0 committed:
compilation fix

CyberROFL commented on ydb/core/tx/schemeshard/schemeshard_export__create.cpp:
```diff
@@ -518,9 +518,9 @@ struct TSchemeShard::TExport::TTxProgress: public TSchemeShard::TXxport::TTxBase
-        if (auto schemeUploader = std::exchange(item.SchemeUploader, {})) {
-            Send(schemeUploader, new TEvents::TEvPoisonPill());
-            Self->RunningExportSchemeUploaders.erase(schemeUploader);
+        if (item.SchemeUploader) {
+            Send(item.SchemeUploader, new TEvents::TEvPoisonPill());
+            Self->RunningExportSchemeUploaders.erase(std::exchange(item.SchemeUploader, TActorId()));
```
Чем предыдущая версия кода была хуже? Тут и так приличный дифф, давай не раздувать его такими nop изменениями.

jepett0 replied:
хотел подчеркнуть всё-таки, что мы делаем что-то только с недефолтным item.SchemeUploader. А сюда залез, чтобы было одинаково с import__create

CyberROFL commented on ydb/core/tx/schemeshard/schemeshard_import__create.cpp:
```diff
@@ -481,6 +576,17 @@ struct TSchemeShard::TImport::TTxProgress: public TSchemeShard::TXxport::TTxBase
     }
 
+    void KillChildActors(TImportInfo::TItem& item) {
+        if (item.SchemeGetter) {
+            Send(item.SchemeGetter, new TEvents::TEvPoisonPill());
+            Self->RunningImportSchemeGetters.erase(std::exchange(item.SchemeGetter, TActorId()));
+        }
```
Как по мне, так, как написано в export__create.cpp — более универсально, потому что легко переживает добавление новых строчек, использующих `SchemeGetter`. А тут, если будет что-то добавлено в конец, придется  переписывать предыдущую строку (с `std::exchange`).

jepett0 replied:
ok

jepett0 committed:
review fixes: std::exchange in if's

pixcc commented on ydb/core/tx/schemeshard/schemeshard_import__create.cpp:
```diff
@@ -338,6 +373,66 @@ struct TSchemeShard::TImport::TTxProgress: public TSchemeShard::TXxport::TTxBase
+                ));
+                Self->RunningImportSchemeQueryExecutors.emplace(item.SchemeQueryExecutor);
+
+                item.State = EState::CreateSchemeObject;
+                item.ViewCreationRetries++;
```
У тебя в `OnSchemeQueryPreparation` переход к `EState::Waiting` происходит только, если `item.ViewCreationRetries == 0`.

В противном случае, импорт кэнцелится. Получается, что если для одной вью надо будет сделать два ретрая, то импорт покэнцелится?

CyberROFL replied:
Сейчас да. Поправим отдельно в https://github.com/ydb-platform/ydb/issues/13932

pnv1 reviewed (approved):
CLI part LGTM
